pkg/proxy/winkernel/proxier.go

```diff
@@ -0,0 +1,1105 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package winkernel
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"reflect"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/helper"
+	apiservice "k8s.io/kubernetes/pkg/api/service"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/proxy/healthcheck"
+	"k8s.io/kubernetes/pkg/util/async"
+)
+
+// KernelCompatTester tests whether the required kernel capabilities are
+// present to run the windows kernel proxier.
+type KernelCompatTester interface {
+	IsCompatible() error
+}
+
+// CanUseWinKernelProxier returns true if we should use the Kernel Proxier
+// instead of the "classic" userspace Proxier.  This is determined by checking
+// the windows kernel version and for the existence of kernel features.
+func CanUseWinKernelProxier(kcompat KernelCompatTester) (bool, error) {
+	// Check that the kernel supports what we need.
+	if err := kcompat.IsCompatible(); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+type WindowsKernelCompatTester struct{}
+
+func (lkct WindowsKernelCompatTester) IsCompatible() error {
+	return nil
+}
+
+type externalIPInfo struct {
+	ip    string
+	hnsID string
+}
+
+type loadBalancerIngressInfo struct {
+	ip    string
+	hnsID string
+}
+
+// internal struct for string service information
+type serviceInfo struct {
+	clusterIP                net.IP
+	port                     int
+	protocol                 api.Protocol
+	nodePort                 int
+	targetPort               int
+	loadBalancerStatus       api.LoadBalancerStatus
+	sessionAffinityType      api.ServiceAffinity
+	stickyMaxAgeMinutes      int
+	externalIPs              []*externalIPInfo
+	loadBalancerIngressIPs   []*loadBalancerIngressInfo
+	loadBalancerSourceRanges []string
+	onlyNodeLocalEndpoints   bool
+	healthCheckNodePort      int
+	hnsID                    string
+	nodePorthnsID            string
+	policyApplied            bool
+}
+
+type hnsNetworkInfo struct {
+	name string
+	id   string
+}
+
+func Log(v interface{}, message string, level glog.Level) {
+	glog.V(level).Infof("%s, %s", message, spew.Sdump(v))
+}
+
+func LogJson(v interface{}, message string, level glog.Level) {
+	jsonString, err := json.Marshal(v)
+	if err == nil {
+		glog.V(level).Infof("%s, %s", message, string(jsonString))
+	}
+}
+
+// internal struct for endpoints information
+type endpointsInfo struct {
+	ip         string
+	port       uint16
+	isLocal    bool
+	macAddress string
```

Comment:
macAddress does not appear to be used anywhere.  If it's always going to be constant, just put that constant into the HNSEndpoint{} creation instead of carrying it around in the endpointsInfo.

Reply:
It may appear like that for now, but I would like to be this way, so that, when we have info about the macAddress, this can be populated with right data.